File: tree.cpp
#include "common.h"

using namespace std;

class btree{
    public:
        btree(int val):data(val), left(nullptr), right(nullptr){};

        void add(int val){
            if(val < data){
                if(left)
                    left->add(val);
                else
                    left = new btree(val);
            return;
            }
            if(right)
                right->add(val);
            else
                right = new btree(val);
        }

        void printTree(btree* ptr){
            if(!ptr) return;
            printTree(ptr->left);
            cout << *ptr << endl;
            printTree(ptr->right);
        }

        int getLeft(btree &ptr){
            if(ptr.left)
                return ptr.left->data;
            return 0;
        }
        
        int getRight(btree &ptr){
            if(ptr.right)
                return ptr.right->data;
            return 0;
        }

        int getData(btree &ptr){
            return ptr.data;
        }

        friend ostream &operator<<(ostream& oStream,btree &p){
            oStream << p.data << " Left: "<< p.getLeft(p) << " Right: "<< p.getRight(p);
            return oStream;
        }
        
    private:
        int data;
        btree* left;
        btree* right;
};

int main(){
    btree root(6);
    root.add(4);
    root.add(5);
    root.add(1);
    root.add(8);
    root.add(7);
    root.add(11);
    root.add(13);
    root.add(9);
    btree* node = &root;
    root.printTree(node);
    return 0;
}

File: tree.py
class tree:
    def __init__(self,data,left=None,right=None):
        self.data = data
        self.left = left
        self.right = right

    def insert(self, dat):
        if self.data > dat:
            if not self.left:
                self.left=tree(dat)
            else:
                self.left.insert(dat)
        else:
            if not self.right:
                self.right=tree(dat)
            else:
                self.right.insert(dat)

    def printTree(self):
        if self.left != None:
            self.left.printTree()
        print(self.data)
        if self.right != None:
            self.right.printTree()
    
    def find(self, p):
        if self.data == p:
            print ("Found",self.data)
            return 
        elif self.data > p:
            if not self.left:
                print ("Not found")
                return
            self.left.find(p)
        else :
            if not self.right:
                print ("Not found")
                return
            self.right.find(p)

b=tree(5)
b.insert(1)
b.insert(4)
b.insert(6)
b.printTree()
b.find(3)
